### ISSUE TITLE
NonEmptyAtclauseDescriptionCheck: Remove @return JavaDoc block tag

### ISSUE THREAD

**Issue by sikhlev** [checkstyle/checkstyle#5387]
`@return` JavaDoc tag should be removed from NonEmptyAtclauseDescriptionCheck because  empty  comment is not allowed by JavaDoc grammar and parser logs the error before this check runs. Taken from #5092 .

**Comment by romani** [comment]
original idea from #5092  was:
> `@return` without any args results in a parse error(javadoc tool) so check doesnt need to check for it.

I see no harm from his validation.
checkstyle and javadoc tool are running at different phases of build, so javadoc tool errors might not appear.
If we report the same violation as javadoc tool it is ok. In far future javadoc tool might do all types of validations , it does not mean that we will need to remove from Checkstyle all that is already covered by javadoc tool.

Side note:
In java we follow strict rule, we do not support execution of checkstyle on non-compiled sources (javac gives errors), as it damage seriously our AST (unpredictable structure), we should not follow blindly the same principle for javadoc .... as there are no grammar of javadoc so we empirically constructed it base on javadoc work.
As soon as javadoc as comment without binding to java code is correct by syntax - checkstyle should work. All validations of javadoc against of java code is area of work of "validator in javadoc tool" and checkstyle.

@Vladlis , fyi.

**Comment by Vladlis** [comment]
@romani ,
>  javadoc tool are running at different phases of build, so javadoc tool errors might not appear

This issue has no deal with javadoc tool. We are talking about an exception produced by Checkstyle javadoc parser, so there is no way `@return` can be validated by `NonEmptyAtclauseDescriptionCheck` - the problem will be caught during parsing. Therefore we don't need `@return` in acceptable tokens at all.

**Comment by pbludov** [comment]
Perhaps the javadoc parser should be changed to allow empty tags "author", "return", "since" and "version"?

What is the benefit of throwing an exception in case of missing description?
The error message is:
```
Exception in thread "main" java.lang.IllegalArgumentException: [ERROR:13] Javadoc comment at column 0 has parse error. Details: no viable alternative at input '     *' while parsing JAVADOC_TAG
```
This is not informative. And almost useless.

**Comment by romani** [comment]
@pbludov , please provide input file, whole copy-paste of terminal output is the best way to share all context to reproduce on another side.

**Comment by pbludov** [comment]
```
interface Test {

    /**
     * @author
     * @deprecated
     * @exception
     * @param
     * @return
     * @see
     * @serial
     * @serialData
     * @serialField
     * @since
     * @throws
     * @version
     * {@code}
     * {@docRoot}
     * {@inheritDoc}
     * {@link}
     * {@linkplain}
     * {@literal}
     * {@value}
     */
    int method() throws Exception;

}
```

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="NonEmptyAtclauseDescription"/>
  </module>
</module>
```

$ java -jar checkstyle-8.5-all.jar -c TestConfig.xml Test.java 
```
Starting audit...
[ERROR] Test.java:4: Javadoc comment at column 0 has parse error. Details: no viable alternative at input '   *' while parsing JAVADOC_TAG [NonEmptyAtclauseDescription]
Audit done.
Checkstyle ends with 1 errors.
```

I would like to see a violation here, not an exception. The plugin for Idea, for example, ignores all exceptions. The code in the development process is often incomplete and causes exceptions. So hiding them for such cases is normal.

If the parser will accept such comments, the checks can later detect and show _all of them_ to the user. In a convenient and detailed form.

**Comment by Vladlis** [comment]
@romani , should we approve the issue?

**Comment by romani** [comment]
> I would like to see a violation here, not an exception.

as there are no official javadoc grammar, we do not know what is correct javadoc, and we tried to look at behavior of javadoc tool to make a decision.

**Comment by romani** [comment]
currently ... as exception happen before Check come to action ... validation of empty description `@return` in Check useless.

Resolution:
it is ok to remove it.
But message we generate to user is terrible. Looks like we need to add one more sentence to guide user to review specification and run javadoc tool for more detailed error message.

Additionally I would do investigation on ability of javadoc tool to generate HTML with ignoring of all errors.
review our grammar and allow empty

**Comment by rnveach** [comment]
@romani 
> we can update grammar to pass such cases too 

empty `@return` use to not cause a violation.
It was removed in Issue https://github.com/checkstyle/checkstyle/issues/4942.

If we implement this Issue and decide to reverse `@return` throwing an exception on missing description, then this issue will have to be reversed as well.
I suggest we do investigation first and make firm decision before continuing any further.

**Comment by romani** [comment]
>empty return use to not cause a violation.
It was removed in Issue #4942 as part of GSoC.

Thanks for reference. Looks like it was my bad to approve that issue #4942 . Result error is terrible and we need to return back to original problem of why #4942 was opened.  Is no good reason, fix for #4942 should be reverted probably.

>We should only mark issue approved once we have done the investigation.

This issue just reflect current status of code, so it is to be approved. There is no way now to validate empty return by Check.

**Comment by rnveach** [comment]
> It could pass many months before we come to rethinking 

And it hurts nothing right now by leaving it in, which is why I think we shouldn't merge this fix until we are sure we are going to go the route of exception.

**Comment by romani** [comment]
Ok, lets come to original problem which is raised this issue, @sikhlev why you open this issue ?

**Comment by rnveach** [comment]
@romani 
> @sikhlev why you open this issue ?

Like discussed in first post, this issue was split from #5092 which was created by @ps-sp at your request in the PR https://github.com/checkstyle/checkstyle/pull/5000#discussion_r135173856 which is what we are now questioning as good or bad.
@sikhlev only got involved because he/she was going to work on #5092 before it was decided to split it apart into 2 issues.

**Comment by rnveach** [comment]
https://github.com/checkstyle/checkstyle/issues/5387#issuecomment-354116546
> The error message is:
This is not informative. And almost useless.

This was moved to a new issue. https://github.com/checkstyle/checkstyle/issues/5443

**Comment by romani** [comment]
@rnveach , looks like we lost connection to @sikhlev .

If we merge this fix, it will cause minor "Breaking backward compatibility". it will just make Check a bit more correct from configuration, but there will be no benefit to users at all. So we cause harm for them to upgrade config and give nothing useful instead.

As we are not use on our ability to report error to user in better way - #5443 . And we do not know in what it will result in....  Lets do not change NonEmptyAtclauseDescriptionCheck for now. As #5443 is fixed we will rethink this is smb come up with issue.

**Comment by romani** [comment]
@rnveach , lets make decision on issue, it might help us decide on PR status, connection with contributor is lost.

**Comment by rnveach** [comment]
@romani for the issue as a whole, it is my opinion we should avoid exceptions as much as possible.
My IDE doesn't crash when I have an incomplete javadoc, neither does oracle's javadoc util. 
Oracle's javadoc may print warnings and errors on the invalid javadocs, but it still finishes it's process. I see these more as violations in our case then exceptions.
What makes things worse is that checkstyle's execution stops when it hits an exception. Multiple bad javadocs would make an initial scan annoying as you can't see the full list of errors, the exception messages are still vague, and the only way to avoid the exceptions is to skip the file completely. There would be no other suppression method.

IMO, I agree not to change the check, but I think we should revert the javadoc grammar code that requires these descriptions and allow them to be optional.

**Comment by romani** [comment]
> it is my opinion we should avoid exceptions as much as possible.

yes, but here we have nuance.
We made strict grammar to parse more items on grammar lever rather than in each Check or have bunch of util methods. We have the same limitation for java code - we do not support execution on non-compiled sources (sources that follow java grammar). Principle is the same, reason is the same.
This nuance with javadoc is extra problem for plugins. As plugins now support execution of checkstyle only after compilation (I even remember some discussion about IDEA plugin, where plugin is executed on non-compiled sources and there was a problem to prevent it, as IDEA do auto save of java file).
Plugins have to adjust behavior and expect exceptions and convert them to violations.

> My IDE doesn't crash when I have an incomplete javadoc, neither does oracle's javadoc util. .... I see these more as violations in our case then exceptions.

probably we should have option to convert such exceptions to violations. But this is another issue.

> What makes things worse is that checkstyle's execution stops when it hits an exception.

plugins have option to not halt on exception, all of them should use it, as no matter how much we invest in testing, exceptions are unavoidable. Thirdparty Checks are even more unreliable and fragile.
Expect non-exceptional execution of checkstyle is not good for plugins. 

> but I think we should revert the javadoc grammar code that requires these descriptions and allow them to be optional.

More strict grammar will help us a lot when we come to javadoc validations. I more in favor of conversion of such type exception to violation. So we need to decide to delegate this to plugins or do this on our side.
If you agree, please open new issue on this and let discuss it separately. As good starting point we can get review maven and eclipse-cs behavior on missed description.

-----------------------

>  I agree not to change the check

Issue is closed.
